File: src/ts_src/twin/path.ts
import { emit } from '@tauri-apps/api/event'
import * as appDir from '@tauri-apps/api/path';


/// 注意,在Android中以下有部分Dir是无法获取的,并且一但尝试获取会导致白屏... (这个bug,真离谱,真不好找,🌿!)
/// 由于本app专注android,所以直接注释掉了,虽然可以加catch()解决,但是没必要

export const appCacheDirPath = await appDir.appCacheDir();
export const appConfigDirPath = await appDir.appConfigDir();
export const appDataDirPath = await appDir.appDataDir();
export const appLocalDataDirPath = await appDir.appLocalDataDir();
export const appLogDirPath = await appDir.appLogDir();
export const audioDirPath = await appDir.audioDir();
export const cacheDirPath = await appDir.cacheDir();
export const configDirPath = await appDir.configDir();
export const dataDirPath = await appDir.dataDir();
// export const desktopPath = await appDir.desktopDir();
export const documentDirPath = await appDir.documentDir();
export const downloadDirPath = await appDir.downloadDir();
// export const executableDirPath = await appDir.executableDir();
// export const fontDirPath = await appDir.fontDir();
// export const homeDirPath = await appDir.homeDir();
export const localDataDirPath = await appDir.localDataDir();
export const pictureDirPath = await appDir.pictureDir();
export const publicDirPath = await appDir.publicDir();
export const resourceDirPath = await appDir.resourceDir();
// export const runtimeDirPath = await appDir.runtimeDir();
// export const templateDirPath = await appDir.templateDir();
export const videoDirPath = await appDir.videoDir();
//


/**
 * 用于将 本地目录结构 发送给Rust
 *
 * 不要问我为什么不直接使用Rust处理, 因为2.0.0-alpha似乎压根就没有这个功能
 */


export function path_send_rust() {
    const app_all_base_path = {
        app_cache_dir_path      : appCacheDirPath,
        app_config_dir_path     : appConfigDirPath,
        app_data_dir_path       : appDataDirPath,
        app_local_data_dir_path : appLocalDataDirPath,
        app_log_dir_path        : appLogDirPath,
        audio_dir_path          : audioDirPath,
        cache_dir_path          : cacheDirPath,
        config_dir_path         : configDirPath,
        data_dir_path           : dataDirPath,
        // desktop_path            : desktopPath,
        document_dir_path       : documentDirPath,
        download_dir_path       : downloadDirPath,
        // executable_dir_path     : executableDirPath,
        // font_dir_path           : fontDirPath,
        // home_dir_path           : homeDirPath,
        local_data_dir_path     : localDataDirPath,
        picture_dir_path        : pictureDirPath,
        public_dir_path         : publicDirPath,
        resource_dir_path       : resourceDirPath,
        // runtime_dir_path        : runtimeDirPath,
        // template_dir_path       : templateDirPath,
        video_dir_path          : videoDirPath,
    }
    emit('event_modify_path', app_all_base_path)
}

File: src-tauri/src/twin/path.rs
//! # Module Introduction
//! 
//! `Author          ` : wind-ws 1424343223@qq.com  
//! 
//! `Date            ` : 2023-05-16 14:30:45
//! 
//! `LastAuthor      ` : wind-ws 1424343223@qq.com  
//! 
//! `LastEditTime    ` : 2023-05-16 14:31:22
//! 
//! `FilePath        ` : /wind_daily/src-tauri/src/twin/path.rs
//! 
//! ## Description  : 
//! 通过Ts获取基本目录


use std::{path::{PathBuf, Path}};
use serde::Deserialize;


#[derive(Deserialize, Debug )]
pub struct AppAllBasePath {
    app_cache_dir_path : PathBuf,
    app_config_dir_path : PathBuf,
    app_data_dir_path : PathBuf,
    app_local_data_dir_path : PathBuf,
    app_log_dir_path : PathBuf,
    audio_dir_path : PathBuf,
    cache_dir_path : PathBuf,
    config_dir_path : PathBuf,
    data_dir_path : PathBuf,
    // desktop_path : PathBuf,
    document_dir_path : PathBuf,
    download_dir_path : PathBuf,
    // executable_dir_path : PathBuf,
    // font_dir_path : PathBuf,
    // home_dir_path : PathBuf,
    local_data_dir_path : PathBuf,
    picture_dir_path : PathBuf,
    public_dir_path : PathBuf,
    resource_dir_path : PathBuf,
    // runtime_dir_path : PathBuf,
    // template_dir_path : PathBuf,
    video_dir_path : PathBuf,
}

/// 注意的是,[App,Log]在2.0.0版本中将被移除,也就是下个版本,因此 我已经剔除
pub enum BaseDirectory {
    Audio,
    Cache,
    Config,
    Data,
    LocalData,
    // Desktop,
    Document,
    Download,
    // Executable,
    // Font,
    // Home,
    Picture,
    Public,
    // Runtime,
    // Template,
    Video,
    Resource,
    AppConfig,
    AppData,//可能会统一用这个玩意
    AppLocalData,
    AppCache, //应用缓存文件的建议目录路径
    AppLog
}

impl BaseDirectory {
    /// 获取你需要的[Path]
    pub fn get_base_path(self)->&'static Path{
        unsafe {
            match self {
                BaseDirectory::Audio => ALL_BASE_PATH.as_ref().unwrap().audio_dir_path.as_path(),
                BaseDirectory::Cache => ALL_BASE_PATH.as_ref().unwrap().cache_dir_path.as_path(),
                BaseDirectory::Config => ALL_BASE_PATH.as_ref().unwrap().config_dir_path.as_path(),
                BaseDirectory::Data => ALL_BASE_PATH.as_ref().unwrap().data_dir_path.as_path(),
                BaseDirectory::LocalData => ALL_BASE_PATH.as_ref().unwrap().local_data_dir_path.as_path(),
                // BaseDirectory::Desktop => ALL_BASE_PATH.as_ref().unwrap().desktop_path.as_path(),
                BaseDirectory::Document => ALL_BASE_PATH.as_ref().unwrap().document_dir_path.as_path(),
                BaseDirectory::Download => ALL_BASE_PATH.as_ref().unwrap().download_dir_path.as_path(),
                // BaseDirectory::Executable => ALL_BASE_PATH.as_ref().unwrap().executable_dir_path.as_path(),
                // BaseDirectory::Font => ALL_BASE_PATH.as_ref().unwrap().font_dir_path.as_path(),
                // BaseDirectory::Home => ALL_BASE_PATH.as_ref().unwrap().home_dir_path.as_path(),
                BaseDirectory::Picture => ALL_BASE_PATH.as_ref().unwrap().picture_dir_path.as_path(),
                BaseDirectory::Public => ALL_BASE_PATH.as_ref().unwrap().public_dir_path.as_path(),
                // BaseDirectory::Runtime => ALL_BASE_PATH.as_ref().unwrap().runtime_dir_path.as_path(),
                // BaseDirectory::Template => ALL_BASE_PATH.as_ref().unwrap().template_dir_path.as_path(),
                BaseDirectory::Video => ALL_BASE_PATH.as_ref().unwrap().video_dir_path.as_path(),
                BaseDirectory::Resource => ALL_BASE_PATH.as_ref().unwrap().resource_dir_path.as_path(),
                BaseDirectory::AppConfig => ALL_BASE_PATH.as_ref().unwrap().app_config_dir_path.as_path(),
                BaseDirectory::AppData => ALL_BASE_PATH.as_ref().unwrap().app_data_dir_path.as_path(),
                BaseDirectory::AppLocalData => ALL_BASE_PATH.as_ref().unwrap().app_local_data_dir_path.as_path(),
                BaseDirectory::AppCache => ALL_BASE_PATH.as_ref().unwrap().app_cache_dir_path.as_path(),
                BaseDirectory::AppLog => ALL_BASE_PATH.as_ref().unwrap().app_log_dir_path.as_path(),
            }
        }
        // 这path模块的代码,敲起来和搬砖一样,累死了~ 🥱
    }
}

static mut ALL_BASE_PATH:Option<AppAllBasePath> = None;


#[tauri::command]
pub fn init_rust_path(json:AppAllBasePath){
    println!("{json:#?}");    
    unsafe{
        ALL_BASE_PATH = Some(json);
    }
}
